[clang] Add nuw attribute to GEPs

hazzlim opened the pull request:
Add nuw attribute to inbounds GEPs where the expression used to form the GEP is an addition of unsigned indices.

hazzlim committed:
[clang] Add nuw attribute to GEPs

Add nuw attribute to inbounds GEPs where the expression used to form the
GEP is an addition of unsigned indices.

nikic commented on clang/test/CodeGen/2005-01-02-ConstantInits.c:
```diff
@@ -7,13 +7,6 @@
-// CHECK: @i = global i32 4, align 4
-// CHECK: @Arr = global [100 x i32] zeroinitializer, align 16
-// CHECK: @foo2.X = internal global ptr getelementptr (i8, ptr @Arr, i64 196), align 8
-// CHECK: @foo2.i23 = internal global i32 0, align 4
```
Something went wrong with update_cc_test_checks here? It has `--check-globals`, so this was supposed to be checked...

hazzlim replied:
Hmm good point - I think the issue is this file uses `--global-value-regex "@.+"`, which prevents matching because the update script internals actually prefix `"^@"` to the regex (and obviously `"^@@.+"` doesn't match)
https://github.com/llvm/llvm-project/blob/0cff3e85db00b5f425cc4ed0d6921445afa891ca/llvm/utils/UpdateTestChecks/common.py#L2280-L2281

I'm not sure what's best to do here - we can remove `--global-value-regex "@.+"`, which restores CHECK lines for these globals, but that will also add CHECK lines for some attributes because they match the regex on line 2281 there...

We could add `--global-value-regex "[A-Za-z].*"` to capture globals only, not sure if that's too hacky?

nikic replied:
Interesting. So the way this is designed you can't cleanly distinguish globals and attributes.

I think you solution sounds fine.

hazzlim replied:
Yeah, possibly something that could be addressed in the script if desired - not sure how much it comes up.

Ok great, I've updated the `--global-value-regex` argument and fixed up the missing global CHECKs.

nikic reviewed (commented):
Thanks, this looks like the right direction.

hazzlim committed:
Fix preservation of global CHECK lines in ConstantInits.c

update_cc_test_checks.py will prefix the --global-value-regex argument
with "^@" and "^!" when trying to match globals, and so the current
argument of "@.+" will not match globals.

Replace "@.+" with "[A-Za-z].*" to correctly match globals and not
attributes.

efriedma-quic reviewed (approved):
LGTM; the approach here seems appropriate.

(The EmitCheckedInBoundsGEP API could probably use a refactor so we don't have calls to isSignedOverflowDefined() scattered all over the place, but that's a separate issue.)

hazzlim commented:
I plan to land this patch later today, unless there are any objections / further comments.

Thanks for reviewing @nikic @efriedma-quic

vitalybuka commented:
This patch breaks:
https://lab.llvm.org/buildbot/#/builders/25/builds/1952
https://lab.llvm.org/buildbot/#/builders/52/builds/1775

david-arm commented:
> This patch breaks: https://lab.llvm.org/buildbot/#/builders/25/builds/1952 https://lab.llvm.org/buildbot/#/builders/52/builds/1775

From the buildbot run I can see 12 or 13 changes in the build that failed. Just out of curiosity how did you find out it was this patch that caused it?

vitalybuka commented:
> > This patch breaks: https://lab.llvm.org/buildbot/#/builders/25/builds/1952 https://lab.llvm.org/buildbot/#/builders/52/builds/1775
> 
> From the buildbot run I can see 12 or 13 changes in the build that failed. Just out of curiosity how did you find out it was this patch that caused it?

Bisected and reverted locally.

david-arm commented:
> > > This patch breaks: https://lab.llvm.org/buildbot/#/builders/25/builds/1952 https://lab.llvm.org/buildbot/#/builders/52/builds/1775
> > 
> > 
> > From the buildbot run I can see 12 or 13 changes in the build that failed. Just out of curiosity how did you find out it was this patch that caused it?
> 
> Bisected and reverted locally.

Thanks! I was just wondering if it was something obvious from the patch or whether there was a bisecting tool to use in github. But bisecting and reverting locally is the same thing I'd do.

vitalybuka commented:
Maybe the patch introduced some miscompile?
The patch affects only clang, but failing tests are LLVM, but tools compiled with patched clang with UBSAN enabled.

Also there are branches checking sanitizers.

zmodem commented:
We're seeing many test failures in Chromium after this change as well: https://crbug.com/362522336

mikaelholmen commented:
Hi,

Downstream we also saw miscompiles with this patch. I've debugged one such case and I end up with GVN doing things I don't understand.
So with
```
opt -passes=gvn bbi-98517.ll -o - -S
```
on
```
target datalayout = "p:16:16"

define i8 @func(i16 %n) {
entry:
  %numbers = alloca [4 x i8], align 1
  store i8 3, ptr %numbers, align 1

  %sub = add i16 %n, -1
  %arrayidx = getelementptr inbounds nuw [4 x i8], ptr %numbers, i16 0, i16 %sub
  %0 = load i8, ptr %arrayidx, align 1
  ret i8 %0
}
```
we get
```
target datalayout = "p:16:16"

define i8 @func(i16 %n) {
entry:
  %numbers = alloca [4 x i8], align 1
  store i8 3, ptr %numbers, align 1
  %sub = add i16 %n, -1
  %arrayidx = getelementptr inbounds nuw [4 x i8], ptr %numbers, i16 0, i16 %sub
  ret i8 undef
}
```
so it decides that no matter the value of input parameter %n, the return value is undef. I don't understand this.
For %n==1 the return value should be 3?

And this only happens when there is "nuw" on the gep.

nikic commented:
@mikaelholmen Thanks for the reproducer, this makes the issue clear. BasicAA is incorrectly returning NoAlias for the pointers due to https://github.com/llvm/llvm-project/pull/98608.

The issue is that the `add` gets decomposed, but we preserve the nuw flag, effectively making it an `add nuw i16 n, -1`.

hazzlim commented:
> @mikaelholmen Thanks for the reproducer, this makes the issue clear. BasicAA is incorrectly returning NoAlias for the pointers due to #98608.
> 
> The issue is that the `add` gets decomposed, but we preserve the nuw flag, effectively making it an `add nuw i16 n, -1`.

Thanks for identifying this - I'm looking into a fix. Would you prefer I revert #98608 in the meantime @nikic ?

nikic commented:
I don't think we need to revert it. Without this patch it's mostly harmless.

I've been working on a fix for this issue. I think we need to track NUW through LinearExpression using something along these lines: https://gist.github.com/nikic/d6986bb62539bba048ced15dd497a4fb

hazzlim commented:
> I don't think we need to revert it. Without this patch it's mostly harmless.
> 
> I've been working on a fix for this issue. I think we need to track NUW through LinearExpression using something along these lines: https://gist.github.com/nikic/d6986bb62539bba048ced15dd497a4fb

Ah great thank you, that’s what I was thinking - if we then make the NoAlias condition added in #98608 conditional on Offset.IsNUW that should fix things, if I understand correctly. I can get a patch up to do this

TerrorJack commented:
This breaks wasm codegen: https://github.com/llvm/llvm-project/issues/108770
